File: features/merchant/shop_page/src/main/java/com/tokopedia/shop/favourite/view/presenter/ShopFavouriteListPresenter.java
package com.tokopedia.shop.favourite.view.presenter;

import com.tokopedia.abstraction.base.view.presenter.BaseDaggerPresenter;
import com.tokopedia.abstraction.common.network.exception.UserNotLoginException;
import com.tokopedia.shop.common.data.source.cloud.model.ShopInfo;
import com.tokopedia.shop.common.domain.interactor.GetShopInfoUseCase;
import com.tokopedia.shop.favourite.data.source.cloud.model.ShopFavouritePagingList;
import com.tokopedia.shop.favourite.data.source.cloud.model.ShopFavouriteUser;
import com.tokopedia.shop.favourite.domain.interactor.GetShopFavouriteUserUseCase;
import com.tokopedia.shop.favourite.domain.model.ShopFavouriteRequestModel;
import com.tokopedia.shop.favourite.view.listener.ShopFavouriteListView;
import com.tokopedia.shop.favourite.view.mapper.ShopFavouriteViewModelMapper;
import com.tokopedia.shop.page.domain.interactor.ToggleFavouriteShopAndDeleteCacheUseCase;
import com.tokopedia.user.session.UserSessionInterface;

import javax.inject.Inject;

import rx.Subscriber;

/**
 * Created by nathan on 2/6/18.
 */

public class ShopFavouriteListPresenter extends BaseDaggerPresenter<ShopFavouriteListView> {

    private static final int SHOP_FAVOURITE_PER_PAGE = 12;

    private final GetShopFavouriteUserUseCase getShopFavouriteUserUseCase;
    private final ToggleFavouriteShopAndDeleteCacheUseCase toggleFavouriteShopAndDeleteCacheUseCase;
    private final ShopFavouriteViewModelMapper shopFavouriteViewModelMapper;
    private final GetShopInfoUseCase getShopInfoUseCase;
    private final UserSessionInterface userSession;

    @Inject
    public ShopFavouriteListPresenter(GetShopFavouriteUserUseCase getShopFavouriteUserUseCase,
                                      ShopFavouriteViewModelMapper shopFavouriteViewModelMapper,
                                      GetShopInfoUseCase getShopInfoUseCase, UserSessionInterface userSession,
                                      ToggleFavouriteShopAndDeleteCacheUseCase toggleFavouriteShopAndDeleteCacheUseCase) {
        this.getShopFavouriteUserUseCase = getShopFavouriteUserUseCase;
        this.toggleFavouriteShopAndDeleteCacheUseCase = toggleFavouriteShopAndDeleteCacheUseCase;
        this.shopFavouriteViewModelMapper = shopFavouriteViewModelMapper;
        this.getShopInfoUseCase = getShopInfoUseCase;
        this.userSession = userSession;
    }

    public void getShopFavouriteList(String shopId, int page) {
        ShopFavouriteRequestModel shopFavouriteRequestModel = new ShopFavouriteRequestModel();
        shopFavouriteRequestModel.setShopId(shopId);
        shopFavouriteRequestModel.setUserId(userSession.getUserId());
        shopFavouriteRequestModel.setDeviceId(userSession.getDeviceId());
        shopFavouriteRequestModel.setPage(page);
        shopFavouriteRequestModel.setPerPage(SHOP_FAVOURITE_PER_PAGE);
        getShopFavouriteUserUseCase.execute(GetShopFavouriteUserUseCase.createRequestParam(shopFavouriteRequestModel), new Subscriber<ShopFavouritePagingList<ShopFavouriteUser>>() {
            @Override
            public void onCompleted() {

            }

            @Override
            public void onError(Throwable e) {
                if (isViewAttached()) {
                    getView().showGetListError(e);
                }
            }

            @Override
            public void onNext(ShopFavouritePagingList<ShopFavouriteUser> shopFavouriteUserPagingList) {
                getView().renderList(shopFavouriteViewModelMapper.transform(shopFavouriteUserPagingList.getList()),
                        shopFavouriteUserPagingList.getPage() < shopFavouriteUserPagingList.getTotalPage());
            }
        });
    }

    public void toggleFavouriteShop(String shopId) {
        if (!userSession.isLoggedIn()) {
            getView().onErrorToggleFavourite(new UserNotLoginException());
            return;
        }
        toggleFavouriteShopAndDeleteCacheUseCase.execute(
                ToggleFavouriteShopAndDeleteCacheUseCase.createRequestParam(shopId), new Subscriber<Boolean>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        if (isViewAttached()) {
                            getView().onErrorToggleFavourite(e);
                        }
                    }

                    @Override
                    public void onNext(Boolean aBoolean) {
                        getView().onSuccessToggleFavourite(aBoolean);
                    }
                });
    }

    public boolean isMyShop(String shopId) {
        return userSession.getShopId().equals(shopId);
    }

    public boolean isLoggedIn() {
        return userSession.isLoggedIn();
    }

    public void getShopInfo(String shopId) {
        getShopInfoUseCase.execute(GetShopInfoUseCase.createRequestParam(shopId), new Subscriber<ShopInfo>() {
            @Override
            public void onCompleted() {

            }

            @Override
            public void onError(Throwable e) {
                if (isViewAttached()) {
                    getView().showGetListError(e);
                }
            }

            @Override
            public void onNext(ShopInfo shopInfo) {
                getView().onSuccessGetShopInfo(shopInfo);
            }
        });
    }

    @Override
    public void detachView() {
        super.detachView();
        getShopFavouriteUserUseCase.unsubscribe();
        toggleFavouriteShopAndDeleteCacheUseCase.unsubscribe();
        getShopInfoUseCase.unsubscribe();

    }
}

File: features/merchant/shop_page/src/main/java/com/tokopedia/shop/favourite/view/presenter/ShopFavouriteListPresenter.kt
package com.tokopedia.shop.favourite.view.presenter

import com.tokopedia.abstraction.base.view.presenter.BaseDaggerPresenter
import com.tokopedia.network.exception.UserNotLoginException
import com.tokopedia.abstraction.common.dispatcher.CoroutineDispatchers
import com.tokopedia.kotlin.extensions.view.toIntOrZero
import com.tokopedia.shop.common.domain.interactor.GQLGetShopInfoUseCase
import com.tokopedia.shop.common.domain.interactor.GQLGetShopInfoUseCase.Companion.FAVORITE_LIST_SOURCE
import com.tokopedia.shop.favourite.data.pojo.shopfollowerlist.GetShopFollowerListData
import com.tokopedia.shop.favourite.data.pojo.shopfollowerlist.ShopFollowerData
import com.tokopedia.shop.favourite.domain.interactor.GetShopFollowerListUseCase
import com.tokopedia.shop.favourite.view.listener.ShopFavouriteListView
import com.tokopedia.shop.favourite.view.model.ShopFollowerListResultUiModel
import com.tokopedia.shop.favourite.view.model.ShopFollowerUiModel
import com.tokopedia.shop.pageheader.domain.interactor.ToggleFavouriteShopAndDeleteCacheUseCase
import com.tokopedia.user.session.UserSessionInterface
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import rx.Subscriber
import javax.inject.Inject
import kotlin.coroutines.CoroutineContext
import kotlin.math.max

/**
 * Created by nathan on 2/6/18.
 */
class ShopFavouriteListPresenter @Inject
constructor(private val getShopFollowerListUseCase: GetShopFollowerListUseCase,
            private val gqlGetShopInfoUseCase: GQLGetShopInfoUseCase,
            private val userSession: UserSessionInterface,
            private val toggleFavouriteShopAndDeleteCacheUseCase: ToggleFavouriteShopAndDeleteCacheUseCase,
            private val dispatchers: CoroutineDispatchers) : BaseDaggerPresenter<ShopFavouriteListView>(), CoroutineScope {

    //region CoroutineScope
    private val job: Job = SupervisorJob()

    override val coroutineContext: CoroutineContext
        get() = dispatchers.main + job

    //endregion

    val isLoggedIn: Boolean
        get() = userSession.isLoggedIn

    fun getShopFavouriteList(shopId: String, page: Int) {
        val thePage = max(1, page)
        launch {
            try {
                val data = getShopFollowerListUseCase.apply {
                    clearRequest()
                    addRequestWithParam(shopId, thePage)
                }.executeOnBackground().convertToUiModel(thePage)
                view?.renderList(data.followerUiModelList, data.isCanLoadMore)
            } catch (e: Throwable) {
                view?.showGetListError(e)
            }
        }
    }

    fun toggleFavouriteShop(shopId: String) {
        if (!userSession.isLoggedIn) {
            view.onErrorToggleFavourite(UserNotLoginException())
            return
        }
        toggleFavouriteShopAndDeleteCacheUseCase.execute(
                ToggleFavouriteShopAndDeleteCacheUseCase.createRequestParam(shopId), object : Subscriber<Boolean>() {
            override fun onCompleted() {

            }

            override fun onError(e: Throwable) {
                if (isViewAttached) {
                    view.onErrorToggleFavourite(e)
                }
            }

            override fun onNext(aBoolean: Boolean?) {
                view.onSuccessToggleFavourite(aBoolean!!)
            }
        })
    }

    fun isMyShop(shopId: String): Boolean {
        return userSession.shopId == shopId
    }

    fun getShopInfo(shopId: String) {
        gqlGetShopInfoUseCase.params = GQLGetShopInfoUseCase.createParams(listOf(shopId.toIntOrZero()), source = FAVORITE_LIST_SOURCE)
        gqlGetShopInfoUseCase.execute(
                {
                    if(isViewAttached) {
                        view?.onSuccessGetShopInfo(it)
                    }
                },
                {
                    if(isViewAttached) {
                        view.showGetListError(it)
                    }
                }
        )
    }

    override fun detachView() {
        super.detachView()
        toggleFavouriteShopAndDeleteCacheUseCase.unsubscribe()
        gqlGetShopInfoUseCase.cancelJobs()
        job.cancel()
    }

    private fun GetShopFollowerListData.convertToUiModel(currentPage: Int) = shopFollowerList.let { result ->
        ShopFollowerListResultUiModel(
                isCanLoadMore = result.haveNext,
                currentPage = currentPage,
                followerUiModelList = result.shopFollowerData.map { it.convertToUiModel() }
        )
    }

    private fun ShopFollowerData.convertToUiModel(): ShopFollowerUiModel = ShopFollowerUiModel(
            id = followerID,
            name = followerName,
            imageUrl = photo,
            profileUrl = profileURL
    )
}